include/ceed-hash.h

```diff
@@ -0,0 +1,149 @@
+/// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+/// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+/// reserved. See files LICENSE and NOTICE for details.
+///
+/// This file is part of CEED, a collection of benchmarks, miniapps, software
+/// libraries and APIs for efficient high-order finite element and spectral
+/// element discretizations for exascale applications. For more information and
+/// source code availability see http://github.com/ceed.
+///
+/// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+/// a collaborative effort of two U.S. Department of Energy organizations (Office
+/// of Science and the National Nuclear Security Administration) responsible for
+/// the planning and preparation of a capable exascale ecosystem, including
+/// software, applications, hardware, advanced system engineering and early
+/// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// Public header for hashing functionality for libCEED, adapted from PETSc
+#ifndef _ceed_hash_h
+#define _ceed_hash_h
+
+#include <ceed.h>
+#include <khash.h>
+
+/* Required for khash <= 0.2.5 */
+#if !defined(kcalloc)
+#define kcalloc(N,Z) calloc(N,Z)
+#endif
+#if !defined(kmalloc)
+#define kmalloc(Z) malloc(Z)
+#endif
+#if !defined(krealloc)
+#define krealloc(P,Z) realloc(P,Z)
+#endif
+#if !defined(kfree)
+#define kfree(P) free(P)
+#endif
+
+/* --- Useful extensions to khash --- */
+
+#if !defined(kh_reset)
+/*! @function
+  @abstract     Reset a hash table to initial state.
+  @param  name  Name of the hash table [symbol]
+  @param  h     Pointer to the hash table [khash_t(name)*]
+ */
+#define kh_reset(name, h) {                                     \
+        if (h) {                                                \
+                kfree((h)->keys); kfree((h)->flags);            \
+                kfree((h)->vals);                               \
+                memset((h), 0x00, sizeof(*(h)));                \
+        } }
+#endif /*kh_reset*/
+
+#if !defined(kh_foreach)
```

Comment:
Why are these here? They're already defined in `khash.h`?

Reply:
I cut out everything we aren't currently using. No clue why they were in the PETSc version of this file.